library/src/main/java/org/xmtp/android/library/ApiClient.kt

```diff
@@ -153,47 +122,125 @@ data class GRPCApiClient(
     override suspend fun batchQuery(
         requests: List<QueryRequest>,
     ): BatchQueryResponse {
-        val batchRequest = BatchQueryRequest.newBuilder().addAllRequests(requests).build()
-        val headers = Metadata()
+        val batchRequest = requests.map { queryRequestToFFi(it) }
+        return batchQueryResponseFromFFi(rustV2Client.batchQuery(FfiV2BatchQueryRequest(requests = batchRequest)))
+    }
 
-        authToken?.let { token ->
-            headers.put(AUTHORIZATION_HEADER_KEY, "Bearer $token")
-        }
-        headers.put(CLIENT_VERSION_HEADER_KEY, Constants.VERSION)
-        if (appVersion != null) {
-            headers.put(APP_VERSION_HEADER_KEY, appVersion)
+    override suspend fun publish(envelopes: List<Envelope>) {
+        val ffiEnvelopes = envelopes.map { envelopeToFFi(it) }
+        val request = FfiPublishRequest(envelopes = ffiEnvelopes)
+
+        rustV2Client.publish(request = request, authToken = authToken ?: "")
+    }
+
+    override suspend fun subscribe(topics: List<String>): Flow<Envelope> = flow {
+        try {
+            val subscription = rustV2Client.subscribe(FfiV2SubscribeRequest(topics))
```

Comment:
I toyed with the idea of spinning up a Android GRPC channel just for the subscribe stuff since it feels better than this rust subscribe work. But thought it would be better to just remove the dependency on android grpc entirely. Curious others thoughts.

Reply:
That's a little scary. Have you been able to test this with the example app to see if it's the tests or the implementation?

Reply:
Ya I ran into this with Groups as well. Everything seems to work when not using the turbine coroutine test methods. But it does make me desire a beta branch for changes like this.

Reply:
And yes the example app works with streaming.